### PR TITLE
[SPARK-20087] Attach accumulators / metrics to 'TaskKilled' end reason

### PR DESCRIPTION
**Pull request by noodle-fb**
## What changes were proposed in this pull request?

The ultimate goal is for listeners to `onTaskEnd` to receive metrics when a task is killed intentionally, since the data is currently just thrown away. This is already done for ExceptionFailure, so this just copies the same approach.

## How was this patch tested?

The unit test in DAGSchedulerSuite that tests this for ExceptionFailure was modified to test the same thing for TaskKilled. I also re-tested all the unit tests modified by the [last change to TaskKilled](https://github.com/apache/spark/commit/8e558041aa0c41ba9fb2ce242daaf6d6ed4d85b7), and made sure they all still pass.

For integration tests, I ran a query that caused a speculative task retry on our deployment, and verified that the metrics showed up in our logging for that retry when it was killed.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Hi @noodle-fb, it seems not a trivial change that does not need a JIRA. Could we create a JIRA and put this in the title (see http://spark.apache.org/contributing.html)?

**Comment by noodle-fb** [comment]
@HyukjinKwon edited with Jira tag, didn't realize that was the naming convention

**Comment by noodle-fb** [comment]
@JoshRosen ping? not sure how to github correctly

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
add to whitelist

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
ok to test

**Comment by jiangxb1987** [comment]
@noodle-fb could you rebase this so we can review it? Thanks!

**Comment by advancedxy** [comment]
@noodle-fb are you still working on this? If not, I may work on it based on your current impl.

I am facing same issue here. The accumulator updates are lost for killed tasks.

**Comment by squito** [comment]
@advancedxy this has been quiet for a long time, so I suggest you just take it over.  I actually think this is so close to complete that very little would need to be done, and credit would most likely go to @noodle-fb .  That said, we may need a little more input on whether or not this is desirable, as it will change the meaning of the aggregated metrics.

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala [hunk 1]
```diff
@@ -212,9 +212,19 @@ case object TaskResultLost extends TaskFailedReason {
...
+case class TaskKilled(
+    reason: String,
+    accumUpdates: Seq[AccumulableInfo] = Seq.empty,
+    private[spark] var accums: Seq[AccumulatorV2[_, _]] = Nil)
+  extends TaskFailedReason {
+
+  override def toErrorString: String = "TaskKilled ($reason)"
   override def countTowardsTaskFailures: Boolean = false
+
+  private[spark] def withAccums(accums: Seq[AccumulatorV2[_, _]]): TaskKilled = {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: I don't think this method is really necessary at all, you could just pass it in the constructor in the places its used.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 2]
```diff
@@ -429,15 +429,42 @@ private[spark] class Executor(
...
+          // Collect latest accumulator values to report back to the driver
+          val accums: Seq[AccumulatorV2[_, _]] =
+            if (task != null) {
+              task.metrics.setExecutorRunTime(System.currentTimeMillis() - taskStart)
+              task.metrics.setJvmGCTime(computeTotalGcTime() - startGCTime)
+              task.collectAccumulatorUpdates(taskFailed = true)
+            } else {
+              Seq.empty
+            }
+          val accUpdates = accums.map(acc => acc.toInfo(Some(acc.value), None))
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: this should be refactored, and not repeated 3 times.

**Review by squito — CHANGES_REQUESTED** [review]

**Comment by noodle-fb** [comment]
@advancedxy, feel free to take this over! @squito, as I remember this, it seemed inconsistent to count metrics for tasks that fail, but not tasks that were killed, machines are doing work in either case. But others might interpret the metrics differently.

**Comment by advancedxy** [comment]
All right then, I will take it over. Of course the credit should go to @noodle-fb.

We can discuss whether this behaviour is desirable or not in the JIRA or the new PR.

**Pull request closed without merging**